nexus/blueprint-execution/src/dns.rs

```diff
@@ -0,0 +1,744 @@
+        //    to appear in the representative system that we're working with.
+        //
+        // 4. Our out-of-service zone does *not* appear in the DNS config,
+        //    neither with an AAAA record nor in an SRV record.
+        //
+        // Together, this tells us that we have SRV records for all services,
+        // that those SRV records all point to at least one of the Omicron zones
+        // for that service, and that we correctly ignored zones that were not
+        // in service.
+
+        // To start, we need a mapping from underlay IP to the corresponding
+        // Omicron zone.
+        let mut omicron_zones_by_ip: BTreeMap<_, _> = blueprint
+            .all_omicron_zones()
+            .filter(|(_, zone)| zone.id != out_of_service_id)
+            .map(|(_, zone)| (zone.underlay_address, zone.id))
+            .collect();
+        println!("omicron zones by IP: {:#?}", omicron_zones_by_ip);
+
+        // We also want a mapping from underlay IP to the corresponding switch
+        // zone.  In this case, the value is the Scrimlet's sled id.
+        let mut switch_sleds_by_ip: BTreeMap<_, _> = sleds_by_id
+            .iter()
+            .filter_map(|(sled_id, sled)| {
+                if sled.is_scrimlet {
+                    let sled_subnet = policy.sleds.get(sled_id).unwrap().subnet;
+                    let switch_zone_ip = get_switch_zone_address(sled_subnet);
+                    Some((switch_zone_ip, *sled_id))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // Now go through all the DNS names that have AAAA records and remove
+        // any corresponding Omicron zone.  While doing this, construct a set of
+        // the fully-qualified DNS names (i.e., with the zone name suffix
+        // appended) that had AAAA records.  We'll use this later to make sure
+        // all the SRV records' targets that we find are valid.
+        let mut expected_srv_targets: BTreeSet<_> = BTreeSet::new();
+        for (name, records) in &blueprint_dns_zone.records {
+            let addrs: Vec<_> = records
+                .iter()
+                .filter_map(|dns_record| match dns_record {
+                    DnsRecord::Aaaa(addr) => Some(addr),
+                    _ => None,
+                })
+                .collect();
+            for addr in addrs {
+                if let Some(zone_id) = omicron_zones_by_ip.remove(addr) {
+                    println!(
+                        "IP {} found in DNS corresponds with zone {}",
+                        addr, zone_id
+                    );
+                    expected_srv_targets.insert(format!(
+                        "{}.{}",
+                        name, blueprint_dns_zone.zone_name
+                    ));
+                    continue;
+                }
+
+                if let Some(scrimlet_id) = switch_sleds_by_ip.remove(addr) {
```

Comment:
Just for my own understanding: if we remove the scrimlet ID the first time we get here for a particular address, why don't we trip up later due to switch zones having multiple DNS records (dendrite / mgs / mgd)?

Reply:
Good question!  I believe there are a few different records here:

- one AAAA record for something like `dendrite-$sled_id.control-plane.oxide.internal`
- three SRV records for stuff like `_mgs._tcp.control-plane.oxide.internal`, `_dendrite._tcp.control-plane.oxide.internal`, and `_mgd._tcp.control-plane.oxide.internal`.  These have a `target` field which points at `dendrite-$sled_id.control-plane.oxide.internal`, not the IP address directly.

This loop iterates over just the names having `AAAA` records.  There should be only one of those for each Scrimlet.  It's the next loop that iterates over the `SRV` records and checks those.